File: SiriusQuality/SiriusQuality-ET/src/cpp/SiriusQuality_ET/Penman.cpp
#define _USE_MATH_DEFINES
#include <cmath>
#include <iostream>
#include <vector>
#include <string>
#include <numeric>
#include <algorithm>
#include <array>
#include <map>
#include <set>
#include <tuple>
#include "Penman.h"
using namespace SiriusQuality-ET;
Penman::Penman() {}
double Penman::getspecificHeatCapacityAir() { return this->specificHeatCapacityAir; }
double Penman::getpsychrometricConstant() { return this->psychrometricConstant; }
double Penman::getrhoDensityAir() { return this->rhoDensityAir; }
double Penman::getAlpha() { return this->Alpha; }
double Penman::getlambdaV() { return this->lambdaV; }
void Penman::setspecificHeatCapacityAir(double _specificHeatCapacityAir) { this->specificHeatCapacityAir = _specificHeatCapacityAir; }
void Penman::setpsychrometricConstant(double _psychrometricConstant) { this->psychrometricConstant = _psychrometricConstant; }
void Penman::setrhoDensityAir(double _rhoDensityAir) { this->rhoDensityAir = _rhoDensityAir; }
void Penman::setAlpha(double _Alpha) { this->Alpha = _Alpha; }
void Penman::setlambdaV(double _lambdaV) { this->lambdaV = _lambdaV; }
void Penman::Calculate_Model(EnergyBalanceCompositeState &s, EnergyBalanceCompositeState &s1, EnergyBalanceCompositeRate &r, EnergyBalanceCompositeAuxiliary &a, EnergyBalanceCompositeExogenous &ex)
{
    //- Name: Penman -Version: 1.0, -Time step: 1
    //- Description:
    //            * Title: Penman Model
    //            * Authors: Peter D. Jamieson, Glen S. Francis, Derick R. Wilson, Robert J. Martin
    //            * Reference: https://doi.org/10.1016/0168-1923(94)02214-5
    //            * Institution: New Zealand Institute for Crop and Food Research Ltd.,
    //New Zealand Institute for Crop and Food Research Ltd.,
    //New Zealand Institute for Crop and Food Research Ltd.,
    //New Zealand Institute for Crop and Food Research Ltd.
    //
    //            * ExtendedDescription: It uses Penmann-Monteith method vase on the availability of wind and vapor pressure daily data
    //            * ShortDescription: It uses Penmann-Monteith method vase on the availability of wind and vapor pressure daily data
    //- inputs:
    //            * name: VPDair
    //                          ** description : vapour pressure density
    //                          ** inputtype : variable
    //                          ** variablecategory : auxiliary
    //                          ** datatype : DOUBLE
    //                          ** max : 1000
    //                          ** min : 0
    //                          ** default : 2.19
    //                          ** unit : hPa
    //            * name: specificHeatCapacityAir
    //                          ** description : Specific heat capacity of dry air
    //                          ** inputtype : parameter
    //                          ** parametercategory : constant
    //                          ** datatype : DOUBLE
    //                          ** max : 1
    //                          ** min : 0
    //                          ** default : 0.00101
    //                          ** unit : 
    //            * name: psychrometricConstant
    //                          ** description : psychrometric constant
    //                          ** inputtype : parameter
    //                          ** parametercategory : constant
    //                          ** datatype : DOUBLE
    //                          ** max : 1
    //                          ** min : 0
    //                          ** default : 0.66
    //                          ** unit : 
    //            * name: rhoDensityAir
    //                          ** description : Density of air
    //                          ** inputtype : parameter
    //                          ** parametercategory : constant
    //                          ** datatype : DOUBLE
    //                          ** max : None
    //                          ** min : None
    //                          ** default : 1.225
    //                          ** unit : 
    //            * name: Alpha
    //                          ** description : Priestley-Taylor evapotranspiration proportionality constant
    //                          ** inputtype : parameter
    //                          ** parametercategory : constant
    //                          ** datatype : DOUBLE
    //                          ** max : 100
    //                          ** min : 0
    //                          ** default : 1.5
    //                          ** unit : 
    //            * name: evapoTranspirationPriestlyTaylor
    //                          ** description : evapoTranspiration of Priestly Taylor
    //                          ** inputtype : variable
    //                          ** variablecategory : rate
    //                          ** datatype : DOUBLE
    //                          ** max : 10000
    //                          ** min : 0
    //                          ** default : 449.367
    //                          ** unit : g m-2 d-1
    //            * name: lambdaV
    //                          ** description : latent heat of vaporization of water
    //                          ** inputtype : parameter
    //                          ** parametercategory : constant
    //                          ** datatype : DOUBLE
    //                          ** max : 10
    //                          ** min : 0
    //                          ** default : 2.454
    //                          ** unit : 
    //            * name: hslope
    //                          ** description : the slope of saturated vapor pressure temperature curve at a given temperature
    //                          ** inputtype : variable
    //                          ** variablecategory : auxiliary
    //                          ** datatype : DOUBLE
    //                          ** max : 1000
    //                          ** min : 0
    //                          ** default : 0.584
    //                          ** unit : hPa degC-1
    //            * name: conductance
    //                          ** description : conductance
    //                          ** inputtype : variable
    //                          ** variablecategory : state
    //                          ** datatype : DOUBLE
    //                          ** max : 10000
    //                          ** min : 0
    //                          ** default : 598.685
    //                          ** unit : m d-1
    //- outputs:
    //            * name: evapoTranspirationPenman
    //                          ** description : evapoTranspiration of Penman Monteith
    //                          ** datatype : DOUBLE
    //                          ** variablecategory : rate
    //                          ** max : 5000
    //                          ** min : 0
    //                          ** unit : g m-2 d-1
    double VPDair = a.getVPDair();
    double evapoTranspirationPriestlyTaylor = r.getevapoTranspirationPriestlyTaylor();
    double hslope = a.gethslope();
    double conductance = s.getconductance();
    double evapoTranspirationPenman;
    evapoTranspirationPenman = evapoTranspirationPriestlyTaylor / Alpha + (1000.00 * (rhoDensityAir * specificHeatCapacityAir * VPDair * conductance / (lambdaV * (hslope + psychrometricConstant))));
    r.setevapoTranspirationPenman(evapoTranspirationPenman);
}

File: SiriusQuality/SiriusQuality-ET/src/simplace/SiriusQuality_ET/Penman.java
package net.simplace.sim.components.SiriusQuality-ET;
import  java.io.*;
import  java.util.*;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.time.LocalDateTime;
import net.simplace.sim.model.FWSimComponent;
import net.simplace.sim.util.FWSimVarMap;
import net.simplace.sim.util.FWSimVariable;
import net.simplace.sim.util.FWSimVariable.CONTENT_TYPE;
import net.simplace.sim.util.FWSimVariable.DATA_TYPE;
import org.jdom2.Element;


public class Penman extends FWSimComponent
{
    private FWSimVariable<Double> VPDair;
    private FWSimVariable<Double> specificHeatCapacityAir;
    private FWSimVariable<Double> psychrometricConstant;
    private FWSimVariable<Double> rhoDensityAir;
    private FWSimVariable<Double> Alpha;
    private FWSimVariable<Double> evapoTranspirationPriestlyTaylor;
    private FWSimVariable<Double> lambdaV;
    private FWSimVariable<Double> hslope;
    private FWSimVariable<Double> conductance;
    private FWSimVariable<Double> evapoTranspirationPenman;

    public Penman(String aName, HashMap<String, FWSimVariable<?>> aFieldMap, HashMap<String, String> aInputMap, Element aSimComponentElement, FWSimVarMap aVarMap, int aOrderNumber)
    {
        super(aName, aFieldMap, aInputMap, aSimComponentElement, aVarMap, aOrderNumber);
    }

    public Penman(){
        super();
    }

    @Override
    public HashMap<String, FWSimVariable<?>> createVariables()
    {
        addVariable(FWSimVariable.createSimVariable("VPDair", "vapour pressure density", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"hPa", 0, 1000, 2.19, this));
        addVariable(FWSimVariable.createSimVariable("specificHeatCapacityAir", "Specific heat capacity of dry air", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"", 0, 1, 0.00101, this));
        addVariable(FWSimVariable.createSimVariable("psychrometricConstant", "psychrometric constant", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"", 0, 1, 0.66, this));
        addVariable(FWSimVariable.createSimVariable("rhoDensityAir", "Density of air", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"", None, None, 1.225, this));
        addVariable(FWSimVariable.createSimVariable("Alpha", "Priestley-Taylor evapotranspiration proportionality constant", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"", 0, 100, 1.5, this));
        addVariable(FWSimVariable.createSimVariable("evapoTranspirationPriestlyTaylor", "evapoTranspiration of Priestly Taylor", DATA_TYPE.DOUBLE, CONTENT_TYPE.rate,"g m-2 d-1", 0, 10000, 449.367, this));
        addVariable(FWSimVariable.createSimVariable("lambdaV", "latent heat of vaporization of water", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"", 0, 10, 2.454, this));
        addVariable(FWSimVariable.createSimVariable("hslope", "the slope of saturated vapor pressure temperature curve at a given temperature", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"hPa degC-1", 0, 1000, 0.584, this));
        addVariable(FWSimVariable.createSimVariable("conductance", "conductance", DATA_TYPE.DOUBLE, CONTENT_TYPE.state,"m d-1", 0, 10000, 598.685, this));
        addVariable(FWSimVariable.createSimVariable("evapoTranspirationPenman", "evapoTranspiration of Penman Monteith", DATA_TYPE.DOUBLE, CONTENT_TYPE.rate,"g m-2 d-1", 0, 5000, null, this));

        return iFieldMap;
    }
    @Override
    protected void process()
    {
        double t_VPDair = VPDair.getValue();
        double t_specificHeatCapacityAir = specificHeatCapacityAir.getValue();
        double t_psychrometricConstant = psychrometricConstant.getValue();
        double t_rhoDensityAir = rhoDensityAir.getValue();
        double t_Alpha = Alpha.getValue();
        double t_evapoTranspirationPriestlyTaylor = evapoTranspirationPriestlyTaylor.getValue();
        double t_lambdaV = lambdaV.getValue();
        double t_hslope = hslope.getValue();
        double t_conductance = conductance.getValue();
        double t_evapoTranspirationPenman = evapoTranspirationPenman.getDefault();
        t_evapoTranspirationPenman = t_evapoTranspirationPriestlyTaylor / t_Alpha + (1000.00d * (t_rhoDensityAir * t_specificHeatCapacityAir * t_VPDair * t_conductance / (t_lambdaV * (t_hslope + t_psychrometricConstant))));
        evapoTranspirationPenman.setValue(t_evapoTranspirationPenman, this);
    }

    @Override
    protected void init()
    {
    }
    public HashMap<String, FWSimVariable<?>> fillTestVariables(int aParamIndex, TEST_STATE aDefineOrCheck)
    {
        return iFieldMap;
    }

    @Override
    protected FWSimComponent clone(FWSimVarMap aVarMap)
    {
        return new Penman(iName, iFieldMap, iInputMap, iSimComponentElement, aVarMap, iOrderNumber);
    }
}